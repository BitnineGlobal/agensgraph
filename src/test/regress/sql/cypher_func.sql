--
-- labels()
--



--
-- simple test
--

--
-- prepare
--
drop graph simpletest cascade;
drop graph complextest1 cascade;
drop graph complextest2 cascade;



--
-- set graph path
--
create graph simpletest;
set graph_path=simpletest;

--
-- make vlabels
--          A
--          |
--  B       C
--  |       |
--  ----D----
--
create vlabel a;
create vlabel b;
create vlabel c inherits ( a );
create vlabel d inherits ( b , c );

--
-- make nodes
--
create (n:a {name:'nA'} ) return n;
create (n:b {name:'nB'} ) return n;
create (n:c {name:'nC'} ) return n;
create (n:d {name:'nD'} ) return n;

--
-- label()
--
match (x:a) return x.name, label(x);
match (x:b) return x.name, label(x);
match (x:c) return x.name, label(x);
match (x:d) return x.name, label(x);

--
-- lables()
--
match (x:a) return x.name, labels(x);
match (x:b) return x.name, labels(x);
match (x:c) return x.name, labels(x);
match (x:d) return x.name, labels(x);
match (x:a) return x.name, labels(x)[0];
match (x:b) return x.name, labels(x)[0];
match (x:c) return x.name, labels(x)[0];
match (x:d) return x.name, labels(x)[0];
match (x:a) return x.name, labels(x)[1];
match (x:b) return x.name, labels(x)[1];
match (x:c) return x.name, labels(x)[1];
match (x:d) return x.name, labels(x)[1];
match (x:a) return x.name, labels(x)[2];
match (x:b) return x.name, labels(x)[2];
match (x:c) return x.name, labels(x)[2];
match (x:d) return x.name, labels(x)[2];



--
-- complex test 1
--

--
-- set graph path
--
create graph complextest1;
set graph_path=complextest1;

--
-- make vlabels
--                     A
--                     |
--                     B       C
--                     |       |
-- --------D--------   ----E----   F   ----G
-- |       |       |       |       |   |   |
-- H       I       J       --------K----   |
--         |       |               |       |
--         ----------------L----------------
--
create vlabel a;
create vlabel b inherits( a );
create vlabel c;
create vlabel d;
create vlabel e inherits( b , c );
create vlabel f;
create vlabel g;
create vlabel h inherits( d );
create vlabel i inherits( d );
create vlabel j inherits( d );
create vlabel k inherits( e , f , g );
create vlabel l inherits( i , j , k , g );

--
-- make nodes
--
create (n:a {name:'nA'} ) return n;
create (n:b {name:'nB'} ) return n;
create (n:c {name:'nC'} ) return n;
create (n:d {name:'nD'} ) return n;
create (n:e {name:'nE'} ) return n;
create (n:f {name:'nF'} ) return n;
create (n:g {name:'nG'} ) return n;
create (n:h {name:'nH'} ) return n;
create (n:i {name:'nI'} ) return n;
create (n:j {name:'nJ'} ) return n;
create (n:k {name:'nK'} ) return n;
create (n:l {name:'nL'} ) return n;

--
-- label()
--
match (x:a) return x.name, label(x);
match (x:b) return x.name, label(x);
match (x:c) return x.name, label(x);
match (x:d) return x.name, label(x);
match (x:e) return x.name, label(x);
match (x:f) return x.name, label(x);
match (x:g) return x.name, label(x);
match (x:h) return x.name, label(x);
match (x:i) return x.name, label(x);
match (x:j) return x.name, label(x);
match (x:k) return x.name, label(x);
match (x:l) return x.name, label(x);

--
-- labels()
--
match (x:a) return x.name, labels(x);
match (x:b) return x.name, labels(x);
match (x:c) return x.name, labels(x);
match (x:d) return x.name, labels(x);
match (x:e) return x.name, labels(x);
match (x:f) return x.name, labels(x);
match (x:g) return x.name, labels(x);
match (x:h) return x.name, labels(x);
match (x:i) return x.name, labels(x);
match (x:j) return x.name, labels(x);
match (x:k) return x.name, labels(x);
match (x:l) return x.name, labels(x);
match (x:a) return x.name, labels(x)[0];
match (x:b) return x.name, labels(x)[0];
match (x:c) return x.name, labels(x)[0];
match (x:d) return x.name, labels(x)[0];
match (x:e) return x.name, labels(x)[0];
match (x:f) return x.name, labels(x)[0];
match (x:g) return x.name, labels(x)[0];
match (x:h) return x.name, labels(x)[0];
match (x:i) return x.name, labels(x)[0];
match (x:j) return x.name, labels(x)[0];
match (x:k) return x.name, labels(x)[0];
match (x:l) return x.name, labels(x)[0];
match (x:a) return x.name, labels(x)[1];
match (x:b) return x.name, labels(x)[1];
match (x:c) return x.name, labels(x)[1];
match (x:d) return x.name, labels(x)[1];
match (x:e) return x.name, labels(x)[1];
match (x:f) return x.name, labels(x)[1];
match (x:g) return x.name, labels(x)[1];
match (x:h) return x.name, labels(x)[1];
match (x:i) return x.name, labels(x)[1];
match (x:j) return x.name, labels(x)[1];
match (x:k) return x.name, labels(x)[1];
match (x:l) return x.name, labels(x)[1];
match (x:a) return x.name, labels(x)[2];
match (x:b) return x.name, labels(x)[2];
match (x:c) return x.name, labels(x)[2];
match (x:d) return x.name, labels(x)[2];
match (x:e) return x.name, labels(x)[2];
match (x:f) return x.name, labels(x)[2];
match (x:g) return x.name, labels(x)[2];
match (x:h) return x.name, labels(x)[2];
match (x:i) return x.name, labels(x)[2];
match (x:j) return x.name, labels(x)[2];
match (x:k) return x.name, labels(x)[2];
match (x:l) return x.name, labels(x)[2];
match (x:a) return x.name, labels(x)[3];
match (x:b) return x.name, labels(x)[3];
match (x:c) return x.name, labels(x)[3];
match (x:d) return x.name, labels(x)[3];
match (x:e) return x.name, labels(x)[3];
match (x:f) return x.name, labels(x)[3];
match (x:g) return x.name, labels(x)[3];
match (x:h) return x.name, labels(x)[3];
match (x:i) return x.name, labels(x)[3];
match (x:j) return x.name, labels(x)[3];
match (x:k) return x.name, labels(x)[3];
match (x:l) return x.name, labels(x)[3];
match (x:a) return x.name, labels(x)[4];
match (x:b) return x.name, labels(x)[4];
match (x:c) return x.name, labels(x)[4];
match (x:d) return x.name, labels(x)[4];
match (x:e) return x.name, labels(x)[4];
match (x:f) return x.name, labels(x)[4];
match (x:g) return x.name, labels(x)[4];
match (x:h) return x.name, labels(x)[4];
match (x:i) return x.name, labels(x)[4];
match (x:j) return x.name, labels(x)[4];
match (x:k) return x.name, labels(x)[4];
match (x:l) return x.name, labels(x)[4];
match (x:a) return x.name, labels(x)[5];
match (x:b) return x.name, labels(x)[5];
match (x:c) return x.name, labels(x)[5];
match (x:d) return x.name, labels(x)[5];
match (x:e) return x.name, labels(x)[5];
match (x:f) return x.name, labels(x)[5];
match (x:g) return x.name, labels(x)[5];
match (x:h) return x.name, labels(x)[5];
match (x:i) return x.name, labels(x)[5];
match (x:j) return x.name, labels(x)[5];
match (x:k) return x.name, labels(x)[5];
match (x:l) return x.name, labels(x)[5];
match (x:a) return x.name, labels(x)[6];
match (x:b) return x.name, labels(x)[6];
match (x:c) return x.name, labels(x)[6];
match (x:d) return x.name, labels(x)[6];
match (x:e) return x.name, labels(x)[6];
match (x:f) return x.name, labels(x)[6];
match (x:g) return x.name, labels(x)[6];
match (x:h) return x.name, labels(x)[6];
match (x:i) return x.name, labels(x)[6];
match (x:j) return x.name, labels(x)[6];
match (x:k) return x.name, labels(x)[6];
match (x:l) return x.name, labels(x)[6];
match (x:a) return x.name, labels(x)[7];
match (x:b) return x.name, labels(x)[7];
match (x:c) return x.name, labels(x)[7];
match (x:d) return x.name, labels(x)[7];
match (x:e) return x.name, labels(x)[7];
match (x:f) return x.name, labels(x)[7];
match (x:g) return x.name, labels(x)[7];
match (x:h) return x.name, labels(x)[7];
match (x:i) return x.name, labels(x)[7];
match (x:j) return x.name, labels(x)[7];
match (x:k) return x.name, labels(x)[7];
match (x:l) return x.name, labels(x)[7];
match (x:a) return x.name, labels(x)[8];
match (x:b) return x.name, labels(x)[8];
match (x:c) return x.name, labels(x)[8];
match (x:d) return x.name, labels(x)[8];
match (x:e) return x.name, labels(x)[8];
match (x:f) return x.name, labels(x)[8];
match (x:g) return x.name, labels(x)[8];
match (x:h) return x.name, labels(x)[8];
match (x:i) return x.name, labels(x)[8];
match (x:j) return x.name, labels(x)[8];
match (x:k) return x.name, labels(x)[8];
match (x:l) return x.name, labels(x)[8];
match (x:a) return x.name, labels(x)[9];
match (x:b) return x.name, labels(x)[9];
match (x:c) return x.name, labels(x)[9];
match (x:d) return x.name, labels(x)[9];
match (x:e) return x.name, labels(x)[9];
match (x:f) return x.name, labels(x)[9];
match (x:g) return x.name, labels(x)[9];
match (x:h) return x.name, labels(x)[9];
match (x:i) return x.name, labels(x)[9];
match (x:j) return x.name, labels(x)[9];
match (x:k) return x.name, labels(x)[9];
match (x:l) return x.name, labels(x)[9];
match (x:a) return x.name, labels(x)[10];
match (x:b) return x.name, labels(x)[10];
match (x:c) return x.name, labels(x)[10];
match (x:d) return x.name, labels(x)[10];
match (x:e) return x.name, labels(x)[10];
match (x:f) return x.name, labels(x)[10];
match (x:g) return x.name, labels(x)[10];
match (x:h) return x.name, labels(x)[10];
match (x:i) return x.name, labels(x)[10];
match (x:j) return x.name, labels(x)[10];
match (x:k) return x.name, labels(x)[10];
match (x:l) return x.name, labels(x)[10];



--
-- complex test 2
--

--
-- set graph path
--
create graph complextest2;
set graph_path=complextest2;

--
-- make vlabels
-- ----A------------
-- |   |       B   |
-- |   |       |   |
-- |   ----D----   |
-- |       |       |
-- |       E--   --F
-- |         |   |
-- ----C     --G--
--     |      | |
--     ----H--- ---I
--         |       |
--         ----J----
--

create vlabel a;
create vlabel b;
create vlabel c inherits( a );
create vlabel d inherits( a , b );
create vlabel e inherits( d );
create vlabel f inherits( a );
create vlabel g inherits( e , f );
create vlabel h inherits( c , g );
create vlabel i inherits( g );
create vlabel j inherits( h , i );

--
-- make nodes
--
create (n:a {name:'nA'} ) return n;
create (n:b {name:'nB'} ) return n;
create (n:c {name:'nC'} ) return n;
create (n:d {name:'nD'} ) return n;
create (n:e {name:'nE'} ) return n;
create (n:f {name:'nF'} ) return n;
create (n:g {name:'nG'} ) return n;
create (n:h {name:'nH'} ) return n;
create (n:i {name:'nI'} ) return n;
create (n:j {name:'nJ'} ) return n;

--
-- label()
--
match (x:a) return x.name, label(x);
match (x:b) return x.name, label(x);
match (x:c) return x.name, label(x);
match (x:d) return x.name, label(x);
match (x:e) return x.name, label(x);
match (x:f) return x.name, label(x);
match (x:g) return x.name, label(x);
match (x:h) return x.name, label(x);
match (x:i) return x.name, label(x);
match (x:j) return x.name, label(x);

--
-- labels()
--
match (x:a) return x.name, labels(x);
match (x:b) return x.name, labels(x);
match (x:c) return x.name, labels(x);
match (x:d) return x.name, labels(x);
match (x:e) return x.name, labels(x);
match (x:f) return x.name, labels(x);
match (x:g) return x.name, labels(x);
match (x:h) return x.name, labels(x);
match (x:i) return x.name, labels(x);
match (x:j) return x.name, labels(x);
match (x:a) return x.name, labels(x)[0];
match (x:b) return x.name, labels(x)[0];
match (x:c) return x.name, labels(x)[0];
match (x:d) return x.name, labels(x)[0];
match (x:e) return x.name, labels(x)[0];
match (x:f) return x.name, labels(x)[0];
match (x:g) return x.name, labels(x)[0];
match (x:h) return x.name, labels(x)[0];
match (x:i) return x.name, labels(x)[0];
match (x:j) return x.name, labels(x)[0];
match (x:a) return x.name, labels(x)[1];
match (x:b) return x.name, labels(x)[1];
match (x:c) return x.name, labels(x)[1];
match (x:d) return x.name, labels(x)[1];
match (x:e) return x.name, labels(x)[1];
match (x:f) return x.name, labels(x)[1];
match (x:g) return x.name, labels(x)[1];
match (x:h) return x.name, labels(x)[1];
match (x:i) return x.name, labels(x)[1];
match (x:j) return x.name, labels(x)[1];
match (x:a) return x.name, labels(x)[2];
match (x:b) return x.name, labels(x)[2];
match (x:c) return x.name, labels(x)[2];
match (x:d) return x.name, labels(x)[2];
match (x:e) return x.name, labels(x)[2];
match (x:f) return x.name, labels(x)[2];
match (x:g) return x.name, labels(x)[2];
match (x:h) return x.name, labels(x)[2];
match (x:i) return x.name, labels(x)[2];
match (x:j) return x.name, labels(x)[2];
match (x:a) return x.name, labels(x)[3];
match (x:b) return x.name, labels(x)[3];
match (x:c) return x.name, labels(x)[3];
match (x:d) return x.name, labels(x)[3];
match (x:e) return x.name, labels(x)[3];
match (x:f) return x.name, labels(x)[3];
match (x:g) return x.name, labels(x)[3];
match (x:h) return x.name, labels(x)[3];
match (x:i) return x.name, labels(x)[3];
match (x:j) return x.name, labels(x)[3];
match (x:a) return x.name, labels(x)[4];
match (x:b) return x.name, labels(x)[4];
match (x:c) return x.name, labels(x)[4];
match (x:d) return x.name, labels(x)[4];
match (x:e) return x.name, labels(x)[4];
match (x:f) return x.name, labels(x)[4];
match (x:g) return x.name, labels(x)[4];
match (x:h) return x.name, labels(x)[4];
match (x:i) return x.name, labels(x)[4];
match (x:j) return x.name, labels(x)[4];
match (x:a) return x.name, labels(x)[5];
match (x:b) return x.name, labels(x)[5];
match (x:c) return x.name, labels(x)[5];
match (x:d) return x.name, labels(x)[5];
match (x:e) return x.name, labels(x)[5];
match (x:f) return x.name, labels(x)[5];
match (x:g) return x.name, labels(x)[5];
match (x:h) return x.name, labels(x)[5];
match (x:i) return x.name, labels(x)[5];
match (x:j) return x.name, labels(x)[5];
match (x:a) return x.name, labels(x)[6];
match (x:b) return x.name, labels(x)[6];
match (x:c) return x.name, labels(x)[6];
match (x:d) return x.name, labels(x)[6];
match (x:e) return x.name, labels(x)[6];
match (x:f) return x.name, labels(x)[6];
match (x:g) return x.name, labels(x)[6];
match (x:h) return x.name, labels(x)[6];
match (x:i) return x.name, labels(x)[6];
match (x:j) return x.name, labels(x)[6];
match (x:a) return x.name, labels(x)[7];
match (x:b) return x.name, labels(x)[7];
match (x:c) return x.name, labels(x)[7];
match (x:d) return x.name, labels(x)[7];
match (x:e) return x.name, labels(x)[7];
match (x:f) return x.name, labels(x)[7];
match (x:g) return x.name, labels(x)[7];
match (x:h) return x.name, labels(x)[7];
match (x:i) return x.name, labels(x)[7];
match (x:j) return x.name, labels(x)[7];
match (x:a) return x.name, labels(x)[8];
match (x:b) return x.name, labels(x)[8];
match (x:c) return x.name, labels(x)[8];
match (x:d) return x.name, labels(x)[8];
match (x:e) return x.name, labels(x)[8];
match (x:f) return x.name, labels(x)[8];
match (x:g) return x.name, labels(x)[8];
match (x:h) return x.name, labels(x)[8];
match (x:i) return x.name, labels(x)[8];
match (x:j) return x.name, labels(x)[8];
match (x:a) return x.name, labels(x)[9];
match (x:b) return x.name, labels(x)[9];
match (x:c) return x.name, labels(x)[9];
match (x:d) return x.name, labels(x)[9];
match (x:e) return x.name, labels(x)[9];
match (x:f) return x.name, labels(x)[9];
match (x:g) return x.name, labels(x)[9];
match (x:h) return x.name, labels(x)[9];
match (x:i) return x.name, labels(x)[9];
match (x:j) return x.name, labels(x)[9];



--
-- cleanup
--
drop graph complextest2 cascade;
drop graph complextest1 cascade;
drop graph simpletest cascade;
