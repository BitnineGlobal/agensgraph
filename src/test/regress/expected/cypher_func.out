--
-- vertex_labels()
--
-- prepare
DROP GRAPH IF EXISTS vertex_labels_simple CASCADE;
NOTICE:  graph "vertex_labels_simple" does not exist, skipping
DROP GRAPH IF EXISTS vertex_labels_complex1 CASCADE;
NOTICE:  graph "vertex_labels_complex1" does not exist, skipping
DROP GRAPH IF EXISTS vertex_labels_complex2 CASCADE;
NOTICE:  graph "vertex_labels_complex2" does not exist, skipping
CREATE GRAPH vertex_labels_simple;
CREATE GRAPH vertex_labels_complex1;
CREATE GRAPH vertex_labels_complex2;
-- simple test
SET graph_path = vertex_labels_simple;
--         a
--         |
-- b       c
-- |       |
-- +-- d --+
CREATE VLABEL a;
CREATE VLABEL b;
CREATE VLABEL c INHERITS (a);
CREATE VLABEL d INHERITS (b, c);
CREATE (:a {name: 'a'});
CREATE (:b {name: 'b'});
CREATE (:c {name: 'c'});
CREATE (:d {name: 'd'});
MATCH (n) RETURN n.name, label(n);
 name | label 
------+-------
 "a"  | "a"
 "b"  | "b"
 "c"  | "c"
 "d"  | "d"
(4 rows)

MATCH (n) RETURN n.name, labels(n);
 name |        labels        
------+----------------------
 "a"  | ["a"]
 "b"  | ["b"]
 "c"  | ["c", "a"]
 "d"  | ["d", "b", "c", "a"]
(4 rows)

MATCH (n) RETURN n.name, labels(n)[0];
 name | labels 
------+--------
 "a"  | "a"
 "b"  | "b"
 "c"  | "c"
 "d"  | "d"
(4 rows)

MATCH (n:c) RETURN n.name, labels(n)[1];
 name | labels 
------+--------
 "c"  | "a"
 "d"  | "b"
(2 rows)

MATCH (n:d) RETURN n.name, labels(n)[2], labels(n)[3];
 name | labels | labels 
------+--------+--------
 "d"  | "c"    | "a"
(1 row)

-- complex test 1
SET graph_path = vertex_labels_complex1;
--             a
--             |
--             b       c
--             |       |
-- +-- d --+   +-- e --+   f   +-- g
-- |   |   |       |       |   |   |
-- h   i   j       +------ k --+   |
--     |   |               |       |
--     +---+------ l ------+-------+
CREATE VLABEL a;
CREATE VLABEL b INHERITS (a);
CREATE VLABEL c;
CREATE VLABEL d;
CREATE VLABEL e INHERITS (b, c);
CREATE VLABEL f;
CREATE VLABEL g;
CREATE VLABEL h INHERITS (d);
CREATE VLABEL i INHERITS (d);
CREATE VLABEL j INHERITS (d);
CREATE VLABEL k INHERITS (e, f, g);
CREATE VLABEL l INHERITS (i, j, k, g);
CREATE (:a {name: 'a'});
CREATE (:b {name: 'b'});
CREATE (:c {name: 'c'});
CREATE (:d {name: 'd'});
CREATE (:e {name: 'e'});
CREATE (:f {name: 'f'});
CREATE (:g {name: 'g'});
CREATE (:h {name: 'h'});
CREATE (:i {name: 'i'});
CREATE (:j {name: 'j'});
CREATE (:k {name: 'k'});
CREATE (:l {name: 'l'});
MATCH (n) RETURN n.name, label(n), labels(n);
 name | label |                         labels                          
------+-------+---------------------------------------------------------
 "a"  | "a"   | ["a"]
 "c"  | "c"   | ["c"]
 "d"  | "d"   | ["d"]
 "f"  | "f"   | ["f"]
 "g"  | "g"   | ["g"]
 "b"  | "b"   | ["b", "a"]
 "e"  | "e"   | ["e", "b", "c", "a"]
 "h"  | "h"   | ["h", "d"]
 "i"  | "i"   | ["i", "d"]
 "j"  | "j"   | ["j", "d"]
 "k"  | "k"   | ["k", "e", "f", "g", "b", "c", "a"]
 "l"  | "l"   | ["l", "g", "i", "j", "k", "d", "e", "f", "b", "c", "a"]
(12 rows)

-- complex test 2
SET graph_path = vertex_labels_complex2;
-- +-- a ----------+
-- |   |       b   |
-- |   |       |   |
-- |   +-- d --+   |
-- |       |       |
-- |       e --+-- f
-- |           |
-- +-- c       g
--     |       |
--     +-- h --+-- i
--         |       |
--         +-- j --+
CREATE VLABEL a;
CREATE VLABEL b;
CREATE VLABEL c INHERITS (a);
CREATE VLABEL d INHERITS (a, b);
CREATE VLABEL e INHERITS (d);
CREATE VLABEL f INHERITS (a);
CREATE VLABEL g INHERITS (e, f);
CREATE VLABEL h INHERITS (c, g);
CREATE VLABEL i INHERITS (g);
CREATE VLABEL j INHERITS (h, i);
CREATE (:a {name: 'a'});
CREATE (:b {name: 'b'});
CREATE (:c {name: 'c'});
CREATE (:d {name: 'd'});
CREATE (:e {name: 'e'});
CREATE (:f {name: 'f'});
CREATE (:g {name: 'g'});
CREATE (:h {name: 'h'});
CREATE (:i {name: 'i'});
CREATE (:j {name: 'j'});
MATCH (n) RETURN n.name, label(n), labels(n);
 name | label |                       labels                       
------+-------+----------------------------------------------------
 "a"  | "a"   | ["a"]
 "b"  | "b"   | ["b"]
 "c"  | "c"   | ["c", "a"]
 "d"  | "d"   | ["d", "a", "b"]
 "f"  | "f"   | ["f", "a"]
 "h"  | "h"   | ["h", "c", "g", "a", "e", "f", "d", "b"]
 "e"  | "e"   | ["e", "d", "a", "b"]
 "g"  | "g"   | ["g", "e", "f", "d", "a", "b"]
 "j"  | "j"   | ["j", "h", "i", "c", "g", "a", "e", "f", "d", "b"]
 "i"  | "i"   | ["i", "g", "e", "f", "d", "a", "b"]
(10 rows)

-- cleanup
DROP GRAPH vertex_labels_complex2 CASCADE;
NOTICE:  drop cascades to 13 other objects
DETAIL:  drop cascades to sequence vertex_labels_complex2.ag_label_seq
drop cascades to vlabel ag_vertex
drop cascades to elabel ag_edge
drop cascades to vlabel a
drop cascades to vlabel b
drop cascades to vlabel c
drop cascades to vlabel d
drop cascades to vlabel e
drop cascades to vlabel f
drop cascades to vlabel g
drop cascades to vlabel h
drop cascades to vlabel i
drop cascades to vlabel j
DROP GRAPH vertex_labels_complex1 CASCADE;
NOTICE:  drop cascades to 15 other objects
DETAIL:  drop cascades to sequence vertex_labels_complex1.ag_label_seq
drop cascades to vlabel ag_vertex
drop cascades to elabel ag_edge
drop cascades to vlabel a
drop cascades to vlabel b
drop cascades to vlabel c
drop cascades to vlabel d
drop cascades to vlabel e
drop cascades to vlabel f
drop cascades to vlabel g
drop cascades to vlabel h
drop cascades to vlabel i
drop cascades to vlabel j
drop cascades to vlabel k
drop cascades to vlabel l
DROP GRAPH vertex_labels_simple CASCADE;
NOTICE:  drop cascades to 7 other objects
DETAIL:  drop cascades to sequence vertex_labels_simple.ag_label_seq
drop cascades to vlabel ag_vertex
drop cascades to elabel ag_edge
drop cascades to vlabel a
drop cascades to vlabel b
drop cascades to vlabel c
drop cascades to vlabel d
-- Added test for AG249, use ln() for all log() calls
-- Create initial graph
CREATE GRAPH ag249_log_to_ln;
SET graph_path = ag249_log_to_ln;
CREATE VLABEL numbers;
CREATE (:numbers {string: '10', numeric: 10});
-- These should fail as there is no rule to cast from string to numeric
MATCH (u:numbers) RETURN log(u.string);
ERROR:  cannot cast "10" (jsonb type string) to numeric
MATCH (u:numbers) RETURN ln(u.string);
ERROR:  cannot cast "10" (jsonb type string) to numeric
MATCH (u:numbers) RETURN log10(u.string);
ERROR:  log10(): number is expected but "10"
-- Check that log() == ln() != log10
MATCH (u:numbers) RETURN log(u.numeric), ln(u.numeric), log10(u.numeric);
         ln         |         ln         |       log10        
--------------------+--------------------+--------------------
 2.3025850929940457 | 2.3025850929940457 | 1.0000000000000000
(1 row)

-- Check with a string constant
RETURN log('10'), ln('10'), log10('10');
ERROR:  log10(): number is expected but "10"
-- Check with a numeric constant;
RETURN log(10), ln(10), log10(10);
        ln        |        ln        |       log10        
------------------+------------------+--------------------
 2.30258509299405 | 2.30258509299405 | 1.0000000000000000
(1 row)

-- Check hybrid query
return log10(10), (select log(10));
       log10        | log 
--------------------+-----
 1.0000000000000000 | 1
(1 row)

-- cleanup
DROP GRAPH ag249_log_to_ln CASCADE;
NOTICE:  drop cascades to 4 other objects
DETAIL:  drop cascades to sequence ag249_log_to_ln.ag_label_seq
drop cascades to vlabel ag_vertex
drop cascades to elabel ag_edge
drop cascades to vlabel numbers
-- Added tests for AG222. These tests test the new function
-- get_last_graph_write_stats(). Which is provided to allow
-- access to statistics on the last graph write operation.
CREATE GRAPH ag222;
SET graph_path = ag222;
CREATE VLABEL vertices;
CREATE ELABEL edges;
-- Should return only 2 inserted vertices and 1 inserted edge
create (:vertices {name: 'Boston'})-[:edges]->(:vertices {name: 'Miami'});
SELECT * FROM get_last_graph_write_stats();
 insertedvertices | insertededges | deletedvertices | deletededges | updatedproperties 
------------------+---------------+-----------------+--------------+-------------------
                2 |             1 |               0 |            0 |                 0
(1 row)

-- Should return only 2 updated properties
MATCH (u:vertices) SET u.property = true;
SELECT * FROM get_last_graph_write_stats();
 insertedvertices | insertededges | deletedvertices | deletededges | updatedproperties 
------------------+---------------+-----------------+--------------+-------------------
                0 |             0 |               0 |            0 |                 2
(1 row)

-- Should return only 1 deleted edge
match (u:vertices)-[e:edges]-() delete e;
SELECT * FROM get_last_graph_write_stats();
 insertedvertices | insertededges | deletedvertices | deletededges | updatedproperties 
------------------+---------------+-----------------+--------------+-------------------
                0 |             0 |               0 |            1 |                 0
(1 row)

-- Should return only 2 deleted vertices
match (u) delete u;
SELECT * FROM get_last_graph_write_stats();
 insertedvertices | insertededges | deletedvertices | deletededges | updatedproperties 
------------------+---------------+-----------------+--------------+-------------------
                0 |             0 |               2 |            0 |                 0
(1 row)

-- cleanup
DROP GRAPH ag222 CASCADE;
NOTICE:  drop cascades to 5 other objects
DETAIL:  drop cascades to sequence ag222.ag_label_seq
drop cascades to vlabel ag_vertex
drop cascades to elabel ag_edge
drop cascades to vlabel vertices
drop cascades to elabel edges
-- Added test for AG-283
CREATE GRAPH AG283;
CREATE ({name: 'arc 0', degree: 0});
MATCH (v) RETURN radians(v.degree);
 radians 
---------
 0
(1 row)

-- cleanup
DROP GRAPH AG283 CASCADE;
NOTICE:  drop cascades to 3 other objects
DETAIL:  drop cascades to sequence ag283.ag_label_seq
drop cascades to vlabel ag_vertex
drop cascades to elabel ag_edge
